[FIX] web_editor: send the html editor writes one after another

rdeodoo opened the pull request:
With the following view hierarchy:
   A
   |
   B' (oe_structure_ID view)

Before this fix, when saving multiple templates at once from the same hierarchy
the COW views (website) would not be saved correctly as the first call to
`write` would COW the parent view, then the second concurent call to `write`
would create the child specific view under the generic parent view.
Instead, the generic parent should have been cloned to a specific parent view
and then the child view should be COW under the specific parent.

  A    A'
  |    |
  B'   B*' (This one is the old B', without B' changes in the html editor)

Instead of

  A   A'
      |
      B'

Calling the `write` one after another will do the trick. Indeed, when the
second call is fired, the first call has already triggered the COW behavior.
Thus, the second call correctly retrieves the COW view.

Another issue might still appear, if the parent view write is sent first.
Indeed, the parent view will be COW, thus the specific child view (second write
waiting to be fired) will be deleted to be added on the new specific hierarchy.
Thus, that write will then raise an error since the view does not exist anymore.

task-1934279

rdeodoo commented:
@qsm-odoo As discussed I just changed the 'specific view retrieving' from `_displayResource` to `_saveResources`.
Can you have a final look?
Basically, this PR:
1. Fix the traceback when you save multiple views at once, if the views are from the same view tree (direct child or not) and one of the parent was generic (it might look like a rare case but it actually his the common case of every multi-website flow, oe-structure etc). That was caused by the fact that when saving a parent, it would remove the child to recreate it under the new specific tree.
2. Add a tour that edit multiple views through html editor at once, from the same hierarchy.
3. Add a tour that edit scss file.
4. Fix JS traceback/UI Glitch when you were saving a generic view (it would try to reload that view after the save (ID in url) but that view would not be found as replaced by COW'd view).

It is quite a small diff without the tests.

PS: Don't call robodoo yet, I still have a cherry-pick'd commit from 11.0 in the PR, I am waiting for the forward-port.

qsm-odoo commented on addons/web_editor/static/src/js/widgets/ace.js:
```diff
@@ -562,12 +564,18 @@ var ViewEditor = Widget.extend({
+        var mutex = new concurrency.Mutex();
         _.each(toSave, (function (_toSave, type) {
-            defs = defs.concat(_.map(_toSave, (type === 'xml' ? this._saveView : this._saveSCSS).bind(this)));
+            // Child views first as COW on parent will delete it
```
"Child views first as COW on **a** parent **would** delete **them**" :wink:

JKE-be replied:
cow should not delete, but move view :'(

rdeodoo replied:
Don't worry, it should be done in saas11.3 for the release of stable multi website

qsm-odoo commented on addons/web_editor/static/src/js/widgets/ace.js:
```diff
@@ -562,12 +564,18 @@ var ViewEditor = Widget.extend({
         }).bind(this));
 
-        return $.when.apply($, defs).fail((function (session, error) {
+        return mutex.getUnlockedDef().fail((function (session, error) {
```
Mmh in fact `getUnlockedDef` never fails. To get the same behavior as before, you have to... do stuff difficult to explain by writing  :cry:

JKE-be replied:
Maybe you can try to draw smthing @qsm-odoo  ? 

"one picture is worth a thousand words" ?

rdeodoo replied:
Done, as discussed needed to `hack` the mutex as .fail() can't happen on mutex

qsm-odoo commented on addons/web_editor/static/src/js/widgets/ace.js:
```diff
@@ -360,7 +360,7 @@ var ViewEditor = Widget.extend({
 
         if (this.currentType === 'xml') {
-            this.$viewID.text(_.str.sprintf(_t("Template ID: %s"), this.views[resID].xml_id));
+            this.$viewID.text(_.str.sprintf(_t("Template ID: %s"), this.views[resID].xml_id || this.views[resID].key));
```
Should not it be 'key or xml_id' instead of 'xml_id or key'?

JKE-be replied:
Should not it be 'key' only ?

by default key = xml_id!
why try to handle the specific case of "spcific view" instead to use common field key

rdeodoo replied:
IMHO all 3 solutions will have the same results as if there is a key, xml_id should be the same.
I will only use key then like in most of the website code, since it is now required and it is a 'valeur sure' :yum: .

qsm-odoo commented on addons/web_editor/controllers/main.py:
```diff
@@ -278,7 +278,7 @@ def get_image_info(self, image_id=None, xml_id=None):
         # Related views must be fetched if the user wants the views and/or the style
         views = request.env["ir.ui.view"].get_related_views(key, bundles=bundles)
-        views = views.read(['name', 'id', 'key', 'xml_id', 'arch', 'active', 'inherit_id'])
+        views = views.read(['name', 'id', 'key', 'xml_id', 'arch', 'active', 'inherit_id', 'website_id'])
```
Can't reference 'website_id' in web_editor ?

JKE-be replied:
:postal_horn:

qsm-odoo commented on addons/website/static/src/js/widgets/ace.js:
```diff
@@ -37,9 +38,27 @@ var WebsiteAceEditor = AceEditor.extend({
-            window.location.reload();
-            return $.Deferred();
+            var defs = [];
+            var selectedView = _.findWhere(this.views, {id: this._getSelectedResource()});
```
Should check currentType before doing this

qsm-odoo commented on addons/website/static/src/js/widgets/ace.js:
```diff
@@ -37,9 +38,27 @@ var WebsiteAceEditor = AceEditor.extend({
+                    fields: ['id'],
+                    domain: [['key', '=', selectedView.key], ['website_id', '=', context.website_id]],
+                }).then((function (view) {
+                    if (view[0]) {
```
Is not this always true ?

rdeodoo replied:
There should always be a view returned as we just saved a generic view and we should always find the newly created COW view.
That is a just a check to avoid js crash later when reading the hash if for some reason the search_read does not find the view, in that case `view[0]` will be `undefined` as view will be `[]`.

qsm-odoo commented on addons/website/static/src/js/widgets/ace.js:
```diff
@@ -37,9 +38,27 @@ var WebsiteAceEditor = AceEditor.extend({
+                }).bind(this)));
+            }
+            return $.when.apply($, defs).then((function () {
+                window.location.reload();
```
`return $.Deferred();`

rdeodoo replied:
Done

JKE-be commented on addons/web_editor/static/src/js/widgets/ace.js:
```diff
@@ -562,12 +564,18 @@ var ViewEditor = Widget.extend({
         if (errorFound) return $.Deferred().reject(errorFound);
 
-        var defs = [];
+        var mutex = new concurrency.Mutex();
```
youhuuuuu a mutex ;) 

ca fait quoi du coup ?   ca attend chaque exec avant de lancer le suivant ? ca lance otut en parralelle, et ton getUnlock attend la fin de tous ?

rdeodoo replied:
ca attend chaque exec avant de lancer le suivant, c'est justement ça le fix
C'est un peu comme
```javascript
var chain = $.when();
for (...) {
    chain = chain.then(function () {
        ...
    }
}
 return $.when(chain)
```
_Mais en plus compliqué :laughing:_

rdeodoo commented:
task 1926103

rdeodoo commented:
Thank you for your review. Everything is done.
Luckily the needed commit https://github.com/odoo/odoo/commit/6930f9c3536ff84abd7735b736bc02f1ae16d7ab was ported to 12.0 so this PR is now ready to be merged

rdeodoo commented:
> "Calling the write one after the order will do the trick as next write will be fired after the previous write did the COW job." ?

Replaced by _"Indeed, when the second call is fired, the first call has already triggered the COW behavior.
Thus, the second call correctly retrieves the COW view."_

> "since the view does not exist**-s** anymore"

Done

> "If the template does not have an xml_id but has a key, we display it instead" -> no

Good catch, it was the last changed I made after our talk, did not change the commit message accordingly. Now:
_"Display the template's key instead of its xml_id. The key is basically a duplicate of the xml_id, mainly used in website."_

> "we search for it's newly created specific"

Is it unclear?

> "fixed in PR 30403" ... use the PR link

Done, it was to prevent chatter spam on push/push force on that other PR, forgot on last rebase for review to change it.

> You reference the task 1926103 but is it the right one?

No, just changed it, wrong copy paste.

Thanks

rdeodoo commented:
My english is a bit weirdo I have to admit :bowtie:

rdeodoo commented:
Last english typo _"we search for **its** newly created specific"_ pushed.
This one looks good to be merged.

qsm-odoo commented on addons/website/views/website_templates.xml:
```diff
@@ -57,6 +57,8 @@
 
         <script type="text/javascript" src="/website/static/src/js/menu/navbar.js"/>
+
+        <script type="text/javascript" src="/website/static/src/js/html_editor.test.js"/>
```
There is a "tour" folder :)

rdeodoo replied:
Yup, I placed it there in the first place but it is only in website module that we got a `tour` folder.
Almost everywhere else the tour-test are placed in js root folder.
It is not really a tour but a test (but not a QUnit test in `tests`) so idk, you definitely want it in the tour folder? In website_sale eg ever tour-test are in js root folder.

qsm-odoo replied:
There is already a tour test in that folder ;) But I suppose we can move all the tests in a proper tests folder when we will do task https://www.odoo.com/web#id=1934445&action=327&model=project.task&view_type=form&menu_id=4720, so you can forget about this.

rdeodoo commented:
Done & done. English should now be reverso.net approved!

qsm-odoo commented:
@robodoo r+

robodoo commented:
Because this PR has multiple commits, I need to know how to merge it:

* `merge` to merge directly, using the PR as merge commit message
* `rebase-merge` to rebase and merge, using the PR as merge commit message
* `rebase-ff` to rebase and fast-forward

qsm-odoo commented:
@robodoo rebase-ff

robodoo commented:
Merge method set to rebase and fast-forward

robodoo commented:
Staging failed: ci/runbot (view more at http://runbot.odoo.com/runbot/build/438244)

KangOl commented:
@robodoo retry as I failed my forward-port.

robodoo commented:
Merged, thanks!
